numpy/core/getlimits.py

```diff
@@ -29,32 +29,99 @@ def _fr1(a):
     return a
 
+
 class MachArLike:
     """ Object to simulate MachAr instance """
-
-    def __init__(self,
-                 ftype,
-                 *, eps, epsneg, huge, tiny, ibeta, **kwargs):
-        params = _MACHAR_PARAMS[ftype]
-        float_conv = lambda v: array([v], ftype)
-        float_to_float = lambda v : _fr1(float_conv(v))
-        float_to_str = lambda v: (params['fmt'] % array(_fr0(v)[0], ftype))
-
-        self.title = params['title']
+    def __init__(self, ftype, *, eps, epsneg, huge, tiny, smallest_normal,
+                 ibeta, hardcode_smallest_subnormal=False, **kwargs):
+        self.params = _MACHAR_PARAMS[ftype]
+        self.ftype = ftype
+        self.title = self.params['title']
+        self.hardcode_smallest_subnormal = hardcode_smallest_subnormal
         # Parameter types same as for discovered MachAr object.
-        self.epsilon = self.eps = float_to_float(eps)
-        self.epsneg = float_to_float(epsneg)
-        self.xmax = self.huge = float_to_float(huge)
-        self.xmin = self.tiny = float_to_float(tiny)
-        self.ibeta = params['itype'](ibeta)
+        self.epsilon = self.eps = self.float_to_float(eps)
+        self.epsneg = self.float_to_float(epsneg)
+        self.xmax = self.huge = self.float_to_float(huge)
+        self.xmin = self.tiny = self.float_to_float(tiny)
+        self.ibeta = self.params['itype'](ibeta)
+        self.smallest_normal = self.float_to_float(smallest_normal)
+        self._str_smallest_normal = self.float_to_str(self.smallest_normal)
         self.__dict__.update(kwargs)
         self.precision = int(-log10(self.eps))
-        self.resolution = float_to_float(float_conv(10) ** (-self.precision))
-        self._str_eps = float_to_str(self.eps)
-        self._str_epsneg = float_to_str(self.epsneg)
-        self._str_xmin = float_to_str(self.xmin)
-        self._str_xmax = float_to_str(self.xmax)
-        self._str_resolution = float_to_str(self.resolution)
+        self.resolution = self.float_to_float(
+            self.float_conv(10) ** (-self.precision))
+        self._str_eps = self.float_to_str(self.eps)
+        self._str_epsneg = self.float_to_str(self.epsneg)
+        self._str_xmin = self.float_to_str(self.xmin)
+        self._str_xmax = self.float_to_str(self.xmax)
+        self._str_resolution = self.float_to_str(self.resolution)
+
+    @property
+    def smallest_subnormal(self):
+        """Return the value for the smallest subnormal.
+
+        Returns
+        -------
+        smallest_subnormal : float
+            value for the smallest subnormal
+
+        Warns
+        -----
+        UserWarning
+            If the calculated value for the smallest subnormal is zero.
+        """
+        ld = ntypes.longdouble
+        if self.hardcode_smallest_subnormal:
+            # It is preferred to hardcode this value here because in some
+            # platforms nextafter will return not expected values.
+            value = exp2(ld(-16445))
```

Comment:
How about instead of `self.hardcode_smallest_subnormal` just pass `smallest_subnormal` and if it is `None` you use the nextafter value instead (if you like).

This is the **incorrect** value though, which is probably also why the test fails.

About `with np.errstate(all="ignore"):` I am not sure I understand. That only congtrols floating point errors, not warnings you raise. You can use that "ignore" very liberally in case warnings are given anywhere.

Reply:
There's a method in Numpy to verify the AIX version? I'm seeing that I can call `oslevel` with subprocess to know if the machine is AIX, but I'm not sure if this is the way we want to go.

Reply:
I don't understand.  You are already "hardcoding" it for double-double, which is fine (I don't care if there may be correct double-double implementations that are not AIX).

I am just asking you to move it out of here to where it was before (and fix the value).